Intel GPU oneDNN upstreaming for library compilation

ZhiweiYan-96 opened the pull request:
# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117098



cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

ZhiweiYan-96 committed:
Intel GPU oneDNN upstreaming for library compilation

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"




cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

malfet reviewed (commented):
This PR looks OK, but please add a description explaining what its trying to achieve (upstreaming is a bit vague)

ZhiweiYan-96 commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -18,6 +18,47 @@ IF(NOT MKLDNN_FOUND)
+      set(DNNL_HOST_COMPILER DEFAULT)
+    endif()
+
+    set(MAKE_COMMAND "cmake" "--build" ".")
```
Hi, @malfet  great thanks for the comments. 
The variable name is confusing here, the `MAKE_COMMAND` here is an argument required for `ExternalProject_Add` and it is expected to be used only for oneDNN compilation. Also, it would not override variables in other part. I will modify it to `XPU_DNNL_MAKE_COMMAND` for avoiding confusion.

ZhiweiYan-96 replied:
@malfet Sincere thanks for you comments again, I've change the MAKE_COMMAND to `DNNL_MAKE_COMMAND` for avoid confusion in current commit

ZhiweiYan-96 commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -18,6 +18,47 @@ IF(NOT MKLDNN_FOUND)
+
+  if(USE_XPU) # Build oneDNN GPU library
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      set(DNNL_HOST_COMPILER "g++")
```
Great thanks for the comments, I am changing it.

EikanWang replied:
@malfet , sorry for the late response. We just got the information from the oneDNN community. Currently, oneDNN for XPU only supports `g++` as its explicit host compiler, and they have confirmed that they will support other compilers. By now, I'm wondering if we can leave it as it is and refine the logic when a particular oneDNN version provides this feature.

ZhiweiYan-96 commented:
> This PR looks OK, but please add a description explaining what its trying to achieve (upstreaming is a bit vague)

Thanks for the reminding. I added the motivation in the PR description. This PR is intended to achieve oneDNN compilation in pytorch. Based on this, we could further add operator support that requires oneDNN, like GEMM and Convolution.

jgong5 commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -18,6 +18,47 @@ IF(NOT MKLDNN_FOUND)
   SET(MKLDNN_ROOT "${PROJECT_SOURCE_DIR}/third_party/ideep/mkl-dnn")
+
+  if(USE_XPU) # Build oneDNN GPU library
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
```
What's the problem if we do not add this check?

EikanWang replied:
oneDNN build logic for XPU requires passing the HOST_COMPILER to its cmake, and it is xpu-specific logic.

- `USE_XPU` intends to bypass the logic for CPU
- `CMAKE_CXX_COMPILER_ID` intends to decide which HOST_COMPILER we should pass to oneDNN CMake.

jgong5 replied:
But why can't we use the default logic but have to check "GNU" specifically?

ZhiweiYan-96 replied:
Hi, @jgong5 oneDNN could only parse string `g++` or `/path/to/g++` and `DEFAULT` three options.  When using default options, no explicit host-compiler is passed to icpx, and icpx would use whatever the host compiler it has as a default one.
 
The ideal way for me to pass host-compiler to oneDNN is using `CMAKE_CXX_COMPILER` . But, on a linux system, `CMAKE_CXX_COMPILER` may not the one oneDNN wants, like `/usr/bin/cxx` as I mentioned in comments.

ZhiweiYan-96 replied:
I may need further check on the behavior of  `DEFAULT` options. 
I think what we need is that oneDNN chose a host compiler that is same as `${CMAKE_CXX_COMPILER}`. I am not sure  whether oneDNN has the possibility to use `clang` even if the PyTorch uses a gnu compiler.

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

EikanWang commented:
@malfet , may I know if we have addressed all your comments?

EikanWang commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -18,6 +18,48 @@ IF(NOT MKLDNN_FOUND)
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      set(DNNL_HOST_COMPILER "g++")
+      # g++ is soft linked to /usr/bin/cxx, oneDNN would not treat it as an absolute path
+    else()
+      set(DNNL_HOST_COMPILER DEFAULT)
```
Please remove the default path as its behavior is undefined.

ZhiweiYan-96 replied:
Has modified this part, XPU oneDNN would not build with compiler other than gnu g++, we may remove the restriction after oneDNN community supports others.

xiaolil1 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

EikanWang commented:
@pytorchbot rebase

xiaolil1 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

EikanWang commented:
@ZhiweiYan-96 , please rebase your PR since most runtime PRs have landed.

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

leizhenyuan committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming for library compilation"

# Motivation

As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)





cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

atalman commented:
@EikanWang how do we test this PR is there a test plan for this ? Maybe you can cut and paste the log output of this build ?

EikanWang commented:
> @EikanWang how do we test this PR is there a test plan for this ? Maybe you can cut and paste the log output of this build ?

@atalman , Thanks for the comments, we verified this PR by #117202 of this stack as the XPU GEMM depends on this PR. In parallel, we will paste the build log here. @ZhiweiYan-96 , please help paste the crucial part of the oneDNN build log for XPU.

ZhiweiYan-96 commented:
Hi, @atalman, Thanks for your comments. I grab output from building log as follows, which could help demonstrate the compilation of onednn-gpu lib.

During cmake configuration, xpu part would output 

```
[4/3054] Creating directories for 'xpu_mkldnn_proj'
[5/3054] No download step for 'xpu_mkldnn_proj'
[6/3054] No update step for 'xpu_mkldnn_proj'
[7/3054] No patch step for 'xpu_mkldnn_proj'
[434/3054] Performing configure step for 'xpu_mkldnn_proj'
-- CMAKE_BUILD_TYPE is unset, defaulting to Release
-- The C compiler identification is IntelLLVM 2024.1.0
-- The CXX compiler identification is IntelLLVM
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
...
```

At the compilation stage, xpu mkldnn build would output
```
[2593/3054] Performing build step for 'xpu_mkldnn_proj'
[1/745] Building C object src/common/CMakeFiles/dnnl_common.dir/ittnotify/jitprofiling.c.o
...

[379/745] Building CXX object src/gpu/compute/CMakeFiles/dnnl_gpu_compute.dir/compute_engine.cpp.o
[380/745] Building CXX object src/gpu/CMakeFiles/dnnl_gpu.dir/gpu_deconvolution_list.cpp.o
```
We can see that object for gpu kernel is compiled.

After compilation, the generated object would be further linked into `libdnnl.a`.
```
[745/745] Linking CXX static library src/libdnnl.a
[2594/3054] No install step for 'xpu_mkldnn_proj'
[2595/3054] Completed 'xpu_mkldnn_proj'
```

We also validate the oneDNN library by operator-level ut like conv. We can successfully observe the gpu primitive execution during testing.
```
onednn_verbose,info,gpu,runtime:DPC++
onednn_verbose,primitive,exec,gpu:0,convolution,jit:ir,forward_training,src_f64::blocked:abcd::f0 wei_f64::blocked:abcd::f0 bia_f64::blocked:a::f0 dst_f64::blocked:abcd::f0,attr-scratchpad:user ,alg:convolution_direct,mb2_ic3oc5_ih10oh10kh3sh1dh0ph1_iw10ow10kw3sw1dw0pw1,1.05908
```

atalman commented:
@pytorchmergebot rebase -b main

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

ZhiweiYan-96 commented:
@pytorchbot  merge

ZhiweiYan-96 commented:
@pytorchbot rebase -b main

ZhiweiYan-96 committed:
Update

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge
